zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java

```diff
@@ -808,74 +840,163 @@ public int compare(PathWithStat left, PathWithStat right) {
      * Produces a paginated list of the children of a given path
      * @param path path of node node to list
      * @param stat stat of the node to list
-     * @param watcher an optional watcher to attach to the node. The watcher is added only once when reaching the end of pagination
-     * @param maxReturned maximum number of children to return. Return one more than this number to indicate truncation
+     * @param watcher an optional watcher to attach to the node. The watcher is added only once
+     *                when reaching the end of pagination
+     * @param maxReturned maximum number of children to return.
      * @param minCzxId only return children whose creation zxid equal or greater than minCzxId
      * @param czxIdOffset how many children with zxid == minCzxId to skip (as returned in previous pages)
-     * @return A list of path with stats
+     * @param outputNextPage info to be used for the next page call
+     * @return A list of child names of the given path
      * @throws NoNodeException if the path does not exist
      */
-    public List<PathWithStat> getPaginatedChildren(String path, Stat stat, Watcher watcher, int maxReturned,
-                                                   long minCzxId, long czxIdOffset)
+    public List<String> getPaginatedChildren(String path, Stat stat, Watcher watcher, int maxReturned,
```

Comment:
There is a lot of changes from the logic perspective. For the first version, I would suggest to stay with original patch unless we hit a performance bottle neck. Otherwise, let's minimize the risk.

Reply:
The logic changed is:
1. add logic to pre-compute the packet length to return the max number of children in a page.
2. for the case of getting all children, if all children can be returned in a page, we'll directly return them without sorting.
3. for the 2nd part returning children in a single page, we have to sort the children. The `PriorityQueue` is replaced with an `ArrayList` to further improve the performance. For the step of adding children to the final list, `PriorityQueue` time is `O(NlgN)`, using list is `O(N)`.

Why I consider performance is critical is, the sorting and adding children logic has to be in the synchronized block. It may affect the performance of other read/write. If the paginated API is occasionally used, I would say it is fine. But if we always use the paginated API for all the getChildren operations, I think it is better to make the performance optimal. So in my opinion, at least 1 and 2 are required. For 3, I could still use the priority queue in the original patch. But since we've changed the code, why not just optimize it further (my thought) :)

Reply:
What I am worrying about is this logic is more diverged from the patch what we imported from. And also, if you change the logic, it means we need more tests to verify it.

Reply:
I understand your concern. Tests are added in `GetChildrenPaginatedTest.java` to cover the changes. Anyway, I think if we adopt the patch, we still treat it as the code we develop and ensure it is working fine. 

The gap is, the original patch is not designed for replacing the non-paginated getChildren. It should be fine as an occasional call only when the non-paginated getChildren can not fetch children. But we use this paginated API as regular call for getChildren in Helix. So the optimizations and changes are done for our use case in Helix.